## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data

df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [3]:
# Use the describe function to gather some basic statistics

df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.

df = df.sort_values("client_id", ascending=False)
df.head()

order_id_order_year_df = df.sort_values(
    ["order_id", "order_year"], ascending=True)
order_id_order_year_df.head(15)

df = df.reset_index(drop=True)
df.head()

del df['line_number']
df.head()


first    last            job               phone  \
0  Samantha  Murray  Herpetologist  534.390.2260x56734   
1  Samantha  Murray  Herpetologist  534.390.2260x56734   
2  Samantha  Murray  Herpetologist  534.390.2260x56734   
3  Samantha  Murray  Herpetologist  534.390.2260x56734   
4  Samantha  Murray  Herpetologist  534.390.2260x56734   

                                     email  client_id  order_id  order_date  \
0  murray.samantha.4079@gonzalez-brown.com      99984   9673005  2023-02-27   
1  murray.samantha.4079@gonzalez-brown.com      99984   6821353  2023-02-17   
2  murray.samantha.4079@gonzalez-brown.com      99984   9673005  2023-02-27   
3  murray.samantha.4079@gonzalez-brown.com      99984   8437912  2023-01-25   
4  murray.samantha.4079@gonzalez-brown.com      99984   7810921  2023-01-16   

   order_week  order_year         item_id     category        subcategory  \
0           9        2023  EUX51497-34-6B  consumables  bathroom supplies   
1           7        2023  AEA17352-13-5M    furniture             tables   
2           9        2023  ZAB46262-52-6E    furniture             stands   
3           4        2023  ADB89726-97-4E  consumables   kitchen supplies   
4           3        2023  ABU04430-78-4B  consumables     paper products   

   unit_price  unit_cost  unit_weight   qty  
0      175.21     162.27        13.83   100  
1      125.47      86.77         0.54   139  
2       30.21      19.49         2.62    61  
3        8.67       6.33         2.22    53  
4      101.40      74.52         8.37  6145

In [5]:
# What three item categories had the most entries?
df['category'].value_counts().head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [6]:
# For the category with the most entries,
# which subcategory had the most entries?

df.loc[df['category'] == 'consumables', 'subcategory'].value_counts().head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [7]:
# Which five clients had the most entries in the data?

df['client_id'].value_counts().head(5)


client_id
33615    220
66037    211
46820    209
24741    207
38378    207
Name: count, dtype: int64

In [21]:
# Store the client ids of those top 5 clients in a list.

top_5_clients = list(df['client_id'].value_counts().head(5).index)
print(top_5_clients)

[33615, 66037, 46820, 24741, 38378]


In [24]:
# How many total units (the qty column) did the
# client with the most entries order order?
df.loc[df["client_id"] == 33615, "qty"].sum()




64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [10]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

df['subtotal'] = df['unit_price'] * df['qty']
df.head()


first    last            job               phone  \
0  Samantha  Murray  Herpetologist  534.390.2260x56734   
1  Samantha  Murray  Herpetologist  534.390.2260x56734   
2  Samantha  Murray  Herpetologist  534.390.2260x56734   
3  Samantha  Murray  Herpetologist  534.390.2260x56734   
4  Samantha  Murray  Herpetologist  534.390.2260x56734   

                                     email  client_id  order_id  order_date  \
0  murray.samantha.4079@gonzalez-brown.com      99984   9673005  2023-02-27   
1  murray.samantha.4079@gonzalez-brown.com      99984   6821353  2023-02-17   
2  murray.samantha.4079@gonzalez-brown.com      99984   9673005  2023-02-27   
3  murray.samantha.4079@gonzalez-brown.com      99984   8437912  2023-01-25   
4  murray.samantha.4079@gonzalez-brown.com      99984   7810921  2023-01-16   

   order_week  order_year         item_id     category        subcategory  \
0           9        2023  EUX51497-34-6B  consumables  bathroom supplies   
1           7        2023  AEA17352-13-5M    furniture             tables   
2           9        2023  ZAB46262-52-6E    furniture             stands   
3           4        2023  ADB89726-97-4E  consumables   kitchen supplies   
4           3        2023  ABU04430-78-4B  consumables     paper products   

   unit_price  unit_cost  unit_weight   qty   subtotal  
0      175.21     162.27        13.83   100   17521.00  
1      125.47      86.77         0.54   139   17440.33  
2       30.21      19.49         2.62    61    1842.81  
3        8.67       6.33         2.22    53     459.51  
4      101.40      74.52         8.37  6145  623103.00

In [11]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

df['total_weight'] = df['unit_weight'] * df['qty']

def shipping_price(total_weight):
    if total_weight >  50:
        shipping_price = 7 * total_weight
    else:
        shipping_price = 10 * total_weight
    return shipping_price

df['shipping_price'] = df['total_weight'].apply(shipping_price)

df.head()



first    last            job               phone  \
0  Samantha  Murray  Herpetologist  534.390.2260x56734   
1  Samantha  Murray  Herpetologist  534.390.2260x56734   
2  Samantha  Murray  Herpetologist  534.390.2260x56734   
3  Samantha  Murray  Herpetologist  534.390.2260x56734   
4  Samantha  Murray  Herpetologist  534.390.2260x56734   

                                     email  client_id  order_id  order_date  \
0  murray.samantha.4079@gonzalez-brown.com      99984   9673005  2023-02-27   
1  murray.samantha.4079@gonzalez-brown.com      99984   6821353  2023-02-17   
2  murray.samantha.4079@gonzalez-brown.com      99984   9673005  2023-02-27   
3  murray.samantha.4079@gonzalez-brown.com      99984   8437912  2023-01-25   
4  murray.samantha.4079@gonzalez-brown.com      99984   7810921  2023-01-16   

   order_week  order_year         item_id     category        subcategory  \
0           9        2023  EUX51497-34-6B  consumables  bathroom supplies   
1           7        2023  AEA17352-13-5M    furniture             tables   
2           9        2023  ZAB46262-52-6E    furniture             stands   
3           4        2023  ADB89726-97-4E  consumables   kitchen supplies   
4           3        2023  ABU04430-78-4B  consumables     paper products   

   unit_price  unit_cost  unit_weight   qty   subtotal  total_weight  \
0      175.21     162.27        13.83   100   17521.00       1383.00   
1      125.47      86.77         0.54   139   17440.33         75.06   
2       30.21      19.49         2.62    61    1842.81        159.82   
3        8.67       6.33         2.22    53     459.51        117.66   
4      101.40      74.52         8.37  6145  623103.00      51433.65   

   shipping_price  
0         9681.00  
1          525.42  
2         1118.74  
3          823.62  
4       360035.55

In [12]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

df['totalPrice_noTax'] = df['subtotal'] + df['shipping_price']
df['tax'] = df['totalPrice_noTax'] * 0.0925
df['total_price'] = df['totalPrice_noTax'] + df['tax']
df.head()

first    last            job               phone  \
0  Samantha  Murray  Herpetologist  534.390.2260x56734   
1  Samantha  Murray  Herpetologist  534.390.2260x56734   
2  Samantha  Murray  Herpetologist  534.390.2260x56734   
3  Samantha  Murray  Herpetologist  534.390.2260x56734   
4  Samantha  Murray  Herpetologist  534.390.2260x56734   

                                     email  client_id  order_id  order_date  \
0  murray.samantha.4079@gonzalez-brown.com      99984   9673005  2023-02-27   
1  murray.samantha.4079@gonzalez-brown.com      99984   6821353  2023-02-17   
2  murray.samantha.4079@gonzalez-brown.com      99984   9673005  2023-02-27   
3  murray.samantha.4079@gonzalez-brown.com      99984   8437912  2023-01-25   
4  murray.samantha.4079@gonzalez-brown.com      99984   7810921  2023-01-16   

   order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
0           9        2023  ...     175.21    162.27       13.83   100   
1           7        2023  ...     125.47     86.77        0.54   139   
2           9        2023  ...      30.21     19.49        2.62    61   
3           4        2023  ...       8.67      6.33        2.22    53   
4           3        2023  ...     101.40     74.52        8.37  6145   

    subtotal  total_weight  shipping_price  totalPrice_noTax           tax  \
0   17521.00       1383.00         9681.00          27202.00   2516.185000   
1   17440.33         75.06          525.42          17965.75   1661.831875   
2    1842.81        159.82         1118.74           2961.55    273.943375   
3     459.51        117.66          823.62           1283.13    118.689525   
4  623103.00      51433.65       360035.55         983138.55  90940.315875   

    total_price  
0  2.971819e+04  
1  1.962758e+04  
2  3.235493e+03  
3  1.401820e+03  
4  1.074079e+06  

[5 rows x 23 columns]

In [13]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

df['cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']
df.head()

first    last            job               phone  \
0  Samantha  Murray  Herpetologist  534.390.2260x56734   
1  Samantha  Murray  Herpetologist  534.390.2260x56734   
2  Samantha  Murray  Herpetologist  534.390.2260x56734   
3  Samantha  Murray  Herpetologist  534.390.2260x56734   
4  Samantha  Murray  Herpetologist  534.390.2260x56734   

                                     email  client_id  order_id  order_date  \
0  murray.samantha.4079@gonzalez-brown.com      99984   9673005  2023-02-27   
1  murray.samantha.4079@gonzalez-brown.com      99984   6821353  2023-02-17   
2  murray.samantha.4079@gonzalez-brown.com      99984   9673005  2023-02-27   
3  murray.samantha.4079@gonzalez-brown.com      99984   8437912  2023-01-25   
4  murray.samantha.4079@gonzalez-brown.com      99984   7810921  2023-01-16   

   order_week  order_year  ... unit_cost unit_weight   qty   subtotal  \
0           9        2023  ...    162.27       13.83   100   17521.00   
1           7        2023  ...     86.77        0.54   139   17440.33   
2           9        2023  ...     19.49        2.62    61    1842.81   
3           4        2023  ...      6.33        2.22    53     459.51   
4           3        2023  ...     74.52        8.37  6145  623103.00   

   total_weight  shipping_price  totalPrice_noTax           tax   total_price  \
0       1383.00         9681.00          27202.00   2516.185000  2.971819e+04   
1         75.06          525.42          17965.75   1661.831875  1.962758e+04   
2        159.82         1118.74           2961.55    273.943375  3.235493e+03   
3        117.66          823.62           1283.13    118.689525  1.401820e+03   
4      51433.65       360035.55         983138.55  90940.315875  1.074079e+06   

        cost  
0   25908.00  
1   12586.45  
2    2307.63  
3    1159.11  
4  817960.95  

[5 rows x 24 columns]

In [14]:
# Create a column for the profit of
# each line using line cost and line price

df['profit'] = df['total_price'] - df['cost']
df.head()

first    last            job               phone  \
0  Samantha  Murray  Herpetologist  534.390.2260x56734   
1  Samantha  Murray  Herpetologist  534.390.2260x56734   
2  Samantha  Murray  Herpetologist  534.390.2260x56734   
3  Samantha  Murray  Herpetologist  534.390.2260x56734   
4  Samantha  Murray  Herpetologist  534.390.2260x56734   

                                     email  client_id  order_id  order_date  \
0  murray.samantha.4079@gonzalez-brown.com      99984   9673005  2023-02-27   
1  murray.samantha.4079@gonzalez-brown.com      99984   6821353  2023-02-17   
2  murray.samantha.4079@gonzalez-brown.com      99984   9673005  2023-02-27   
3  murray.samantha.4079@gonzalez-brown.com      99984   8437912  2023-01-25   
4  murray.samantha.4079@gonzalez-brown.com      99984   7810921  2023-01-16   

   order_week  order_year  ... unit_weight   qty   subtotal  total_weight  \
0           9        2023  ...       13.83   100   17521.00       1383.00   
1           7        2023  ...        0.54   139   17440.33         75.06   
2           9        2023  ...        2.62    61    1842.81        159.82   
3           4        2023  ...        2.22    53     459.51        117.66   
4           3        2023  ...        8.37  6145  623103.00      51433.65   

   shipping_price  totalPrice_noTax           tax   total_price       cost  \
0         9681.00          27202.00   2516.185000  2.971819e+04   25908.00   
1          525.42          17965.75   1661.831875  1.962758e+04   12586.45   
2         1118.74           2961.55    273.943375  3.235493e+03    2307.63   
3          823.62           1283.13    118.689525  1.401820e+03    1159.11   
4       360035.55         983138.55  90940.315875  1.074079e+06  817960.95   

          profit  
0    3810.185000  
1    7041.131875  
2     927.863375  
3     242.709525  
4  256117.915875  

[5 rows x 25 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [15]:
# Check your work using the totals above
def total_price_calculation(order_id, df):
    total_price = df[df['order_id'] == order_id]['total_price'].sum()
    return total_price

order_id = [2742071, 2173913, 6128929]
for order_number in order_id:
    total_price = total_price_calculation(order_number, df)
    display(f"Order ID {order_number} has a total price of ${total_price:.2f}")



'Order ID 2742071 has a total price of $152811.90'

'Order ID 2173913 has a total price of $162388.72'

'Order ID 6128929 has a total price of $923441.24'

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [26]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
def client_spent(client_number, total_price):
    client_df = df.loc[df['client_id'] == client_number, total_price]
    return round(client_df.sum(), 2)

for client_number in top_5_clients:
    print(f"{client_number}: ${client_spent(client_number, 'total_price')}")



33615: $8377308.57
66037: $10259514.8
46820: $9743794.32
24741: $82268891.98
38378: $12906550.88


In [17]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

top_5_clients_dict = {"client_number": top_5_clients}


columns = ['qty', 'shipping_price', 'total_price','cost', 'profit']

for summary in columns:
    top_5_clients_dict[summary] = [client_spent(client_number, summary) for client_number in top_5_clients]

summary_df = pd.DataFrame(top_5_clients_dict).sort_values("profit", ascending=True)
summary_df



client_number     qty  shipping_price  total_price         cost  \
0          33615   64313      1828984.89   8377308.57   6175313.91   
2          46820   75768      1601448.84   9743794.32   7007191.64   
1          66037   43018      1395151.85  10259514.80   7004482.98   
4          38378   73667      3429455.40  12906550.88   9634720.98   
3          24741  239862      5126448.37  82268891.98  45688899.71   

        profit  
0   2201994.66  
2   2736602.68  
1   3255031.82  
4   3271829.90  
3  36579992.27

In [18]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

column_dict = {'qty':'Quantity', 'cost': 'Cost', 'shipping_price': 'Shipping_Price', 'total_price': 'Total_Price', 'profit': 'Profit'}
summary_df = summary_df.rename(columns=column_dict)
summary_df

def format_millions(value):
    if value >= 1000000:
        return f"${value/1000000:.2f}M"
    else:
        return f"${value:,.2f}"

summary_df[['Cost', 'Shipping_Price', 'Total_Price', 'Profit']] = summary_df[['Cost', 'Shipping_Price', 'Total_Price', 'Profit']].map(format_millions)
summary_df

            

client_number  Quantity Shipping_Price Total_Price     Cost   Profit
0          33615     64313         $1.83M      $8.38M   $6.18M   $2.20M
2          46820     75768         $1.60M      $9.74M   $7.01M   $2.74M
1          66037     43018         $1.40M     $10.26M   $7.00M   $3.26M
4          38378     73667         $3.43M     $12.91M   $9.63M   $3.27M
3          24741    239862         $5.13M     $82.27M  $45.69M  $36.58M

In [19]:
# Sort the updated data by "Total Profit" form highest to lowest
summary_df.sort_values('Profit', ascending=False)


client_number  Quantity Shipping_Price Total_Price     Cost   Profit
3          24741    239862         $5.13M     $82.27M  $45.69M  $36.58M
4          38378     73667         $3.43M     $12.91M   $9.63M   $3.27M
1          66037     43018         $1.40M     $10.26M   $7.00M   $3.26M
2          46820     75768         $1.60M      $9.74M   $7.01M   $2.74M
0          33615     64313         $1.83M      $8.38M   $6.18M   $2.20M